batch_requests.go

```diff
@@ -0,0 +1,150 @@
+package msgraphgocore
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/google/uuid"
+	abstractions "github.com/microsoft/kiota-abstractions-go"
+	"github.com/microsoftgraph/msgraph-sdk-go-core/internal"
+)
+
+// SendBatch sends a batch request
+func SendBatch(batch batchRequest, adapter abstractions.RequestAdapter) (BatchResponse, error) {
+	var res BatchResponse
+
+	batchJsonBody, err := batch.toJson()
+	if weGotAnError(err) {
+		return res, err
+	}
+
+	baseUrl, err := getBaseUrl(adapter)
+	if weGotAnError(err) {
+		return res, err
+	}
+
+	requestInfo := buildRequestInfo(batchJsonBody, baseUrl)
+	return sendBatchRequest(requestInfo, adapter)
+}
+
+// AppendBatchItem converts RequestInformation to a BatchItem and adds it to a BatchRequest
+//
+// You can add upto 20 BatchItems to a BatchRequest
+func (br *batchRequest) AppendBatchItem(reqInfo abstractions.RequestInformation) (*batchItem, error) {
+	if len(br.Requests) > 19 {
```

Comment:
defensive programming on the request being nil

Reply:
but it should be passed by reference and not by value copy since that struct also has pointers (headers & co, we got burnt by that in the past)

Reply:
A small nitpick, goes is not a pass by reference language. What it does is make a copy of the pointer. I don't understand the scenario you're describing. Do you mind providing a snippet?

Not an expert on defensive programming and reading online suggests the point of defensive programming is to handle unexpected inputs to avoid the program from crashing with errors. I assume you want an error returned in case the value is nil, doesn't that beat the point of defensive programming then?

Reply:
I think we're talking about the same thing, the reference the pointer contains is copied and passed to the callee.

In the early days of the SDK, because request information's headers, path and query parameters are maps on the struct (so pointers), and before the request adapter/authentication layer accepted a pointer to the request information, and before the Uri was a pointer too we've had the something along those lines.

1. new request information
2. pass to request adapter (value copy)
3. pass to the auth layer (value copy)
4. check the URI (set the field in that copy, clear the query parameters for all copies because of pointer)
5. add the authorization header for all copies
6. return from auth
7. try to get the uri again to build a native request in request adapter, but now the URI value is gone (was set on a copy) and the query parameters have been cleared (shared between copies)

I'm worried about similar unanticipated side effects should we work on multiple copies of the object on different layers.